\documentclass[a4paper]{article}

% Support for Norwegian letters
\usepackage[latin1]{inputenc}
\usepackage[T1]{fontenc}
\usepackage{amsmath,amsfonts,amssymb,amsthm,booktabs,array,mathtools}
% consider package mhchem for typesetting chemical formulas

% Proper space and font for integral differential term
\newcommand{\dd}{\; \mathrm{d}} 
% Shorcut for ODEs with proper font
\newcommand{\diff}[2]{\frac{\mathrm{d} #1}{\mathrm{d} #2}}
% Shortcut for PDEs with proper font (shortcut: PDB)
\newcommand{\pdiff}[2]{\frac{\partial #1}{\partial #2}}
\newcommand{\pdiffn}[3]{\frac{\partial^{#3} #1}{\partial #2^{#3}}}

% Absolute value and norm commands.
% Read the mathtools.pdf to fix these!
\providecommand{\abs}[1]{\lvert#1\rvert} 
\providecommand{\norm}[1]{\lVert#1\rVert}

% Set the depth of section numbering
\setcounter{secnumdepth}{0}

\begin{document} 
Synopsis

Gencode has a shitload of cell-lines and compartments. The compartmentalization
si the most interesting one.

Main points

1) Sensitivity (we only get high-rpkm genes) diagram
2) Coverage (Venn diagram)
3) Differences in high RPKM genes
4) Cytoplasm/nucleus differences in polyadenylation (compared to whole cell?)

and 3) % XXX % of these genes have XX poly(A) events and the difference in each
%cell line was III sites per transcript on average
% how many percent of > 1 sites are supported by more than one cell line?

We found signals of polyadenylation events in the GENCODE cell lines by
trimming and remapping those reads that were originally unmappable and ended in
a stretch of As or beginning with a stretch of Ts. We subsequently merged the
polyadenylation sites into clusters in a similar way to Tian et. al
\cite{tian_large-scale_2005}. In total, for all cell lines, we obtained XXX
polyadenylation sites, were XXX \% were common among all cell lines, and XYX
were common to at least two cell lines. To compare our findings with annotated
polyadenylation sites, we first merged and clustered 43187 sites from the
polyAdb with 35791 sites from GENCODE to obtain a total of 50696 annotated
polyA sites. XXX of our polyA sites fall at annotated ones (xxx (\%) of these
are in annotated 3UTRs).

Since the RNA-seq protocol for the Gingeras data is not optimized for 3' ends,
we expected to find most poly(A) sites for transcripts with high RPKM. To
measure this, we calculated the ratio of discovered to annotated poly(A) sites
in the 3UTRs of transcripts with non-overlapping 3UTRs. Figure X shows the
relationship between RPKM and poly(A) discovery ratio. As can be seen, there is
a positive relationship between RPKM and poly(A) discovery (r = 0.52, p <
$10^{-10}$), however there is considerable variation, even for high RPKM
transcripts. We considered an RPKM of minimum 0.5 to count as expressed. For
individual datasets, the average number of poly(A) sites per 3UTR was 1.35, in
agreement with what has been found with \dots , and the average ratio of
poly(A) sites to annotated was 0.75, probably reflecting both that the method
is not exhaustive, and that not all annotated poly(A) sites are used by
transcripts at a given time. When combining all poly(A) sites, the average
number of discovered poly(A) sites per 3UTR was XXXYXY

Of a total XYX clusters in he 3UTRs, X land at annotated
poly(A) sites, while XUU may be novel ones. XYX pcnt of the clusters 3UTRs have
one of the 8 polyadenylation signals within 40 nt downstream (XY for an, and 42
for non-an), also in line with what has been reported previously, showing that
the method is highly specific for detecting polyA sites.

We wanted to compare poly(A) sites in the nucleus and cytoplasm. Poly(A) sites
in the nucleus are expected to originate from stable mRNA that may exist in
multiple copies. The poly(A) sites identified in the nucleus are expected to
stem from mRNA diffusing toward the cytoplasm, mRNA undergoing
processing, and possibly some RNA undergoing degradation
\cite{shcherbik_polyadenylation_2010,slomovic_addition_2010}. We merged the
poly(A) sites from 6 cell lines in the nucleus and cytoplasm, and compared the
distribution of the polyadenylation sites in the genome, as can be seen in
Figure YYXYXY. As can be seen, the nuclear regions capture polyadenylation
signals both in the poly(A)+ and poly(A)- fractions, while in the cytoplasm
most polyadenylation are in the poly(A)+ fraction, and the whole cell extracts
resemple the cytoplasmic ones strongly.

\bibliographystyle{plain}
\bibliography{/users/rg/jskancke/phdproject/bibtex/jorgsk}

\end{document}


